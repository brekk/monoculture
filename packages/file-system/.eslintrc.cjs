module.exports = {
  root: true,
  extends: ['monoculture'],
  ignorePatterns: ['file-system.js'],
}
