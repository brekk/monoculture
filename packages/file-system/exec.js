import { pipe, curry, prop, ifElse, propOr, __ as $ } from 'ramda'
import { execa } from 'execa'
import { Future } from 'fluture'

export const didFail = propOr(true, 'failed')
export const fail = prop('stderr')
/* eslint-disable max-len */
/**
 * @pageSummary Call external processes and consume them as a Future-wrapped value.
 * This is a light wrapper around `execa`, so please see [execa](https://github.com/sindresorhus/execa)'s documentation for more info.
 */

/* eslint-disable jsdoc/tag-lines */
/**
 * Consume external commands as a Future-wrapped value.
 * @curried
 *  1. execWithConfig - Passes all possible configuration values plus a cancellation function.
 *
 *     @example
 *     ```js
 *     import { execWithConfig } from 'file-system'
 *     import { fork } from 'fluture'
 *     fork(console.warn)(console.log)(
 *       execWithConfig(
 *         function customCancellationFunction() {},
 *         'echo',
 *         { cleanup: true },
 *         ['ahoy']
 *       )
 *     )
 *     ```
 *
 *  2. execWithCancel - Eschews any configuration and instead only expects a cancellation function, command and arguments.
 *
 *     @example
 *     ```js
 *     import { execWithCancel } from 'file-system'
 *     import { fork } from 'fluture'
 *     fork(console.warn)(console.log)(
 *       execWithCancel(
 *         function customCancellationFunction() {},
 *         'echo',
 *         ['ahoy']
 *       )
 *     )
 *     ```
 *  3. exec - Eschews any configuration or cancellation function. Needs only command and arguments.
 *
 *     @example
 *     ```js
 *     import { exec } from 'file-system'
 *     import { fork } from 'fluture'
 *     fork(console.warn)(console.log)(
 *       exec(
 *         'echo',
 *         ['ahoy']
 *       )
 *     )
 *     ```
 */
/* eslint-enable max-len */
export const execWithConfig = curry(
  (cancellation, cmd, opts, args) =>
    new Future((bad, good) => {
      execa(cmd, args, opts)
        .catch(pipe(fail, bad))
        .then(ifElse(didFail, pipe(fail, bad), good))
      return cancellation
    })
)
export const execWithCancel = execWithConfig($, $, undefined)
export const exec = execWithCancel(() => {})
