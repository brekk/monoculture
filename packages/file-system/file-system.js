#!/usr/bin/env node

// src/flexeca.js
import { pipe, curry, ifElse, propOr } from "ramda";
import { execa } from "execa";
import { Future } from "fluture";
var didFail = propOr(true, "failed");
var fail = propOr("Something broke", "stderr");
var flexecaWithCanceller = curry(
  (cancellation, cmd, args) => new Future((bad, good) => {
    execa(cmd, args).catch(pipe(fail, (z) => `FAILED: "${z}"`, bad)).then(ifElse(didFail, pipe(fail, bad), good));
    return cancellation;
  })
);
var flexeca = flexecaWithCanceller(() => {
});

// src/fs.js
import {
  rm as __rm,
  mkdir as __mkdir,
  readFile as __readFile,
  writeFile as __writeFile,
  access as __access,
  constants
} from "node:fs";
import { propOr as propOr2, without, curry as curry2, pipe as pipe2, map } from "ramda";
import { chain, chainRej, Future as Future2, parallel } from "fluture";
import glob from "glob";
var localize = (z) => `./${z}`;
var readFile = (x) => new Future2((bad, good) => {
  __readFile(x, "utf8", (err, data) => err ? bad(err) : good(data));
  return () => {
  };
});
var readJSONFile = pipe2(readFile, map(JSON.parse));
var readDirWithConfig = curry2(
  (conf, g) => Future2((bad, good) => {
    try {
      glob(
        g,
        conf,
        (e, x) => (
          // thus far I cannot seem to ever call `bad` from within here
          e ? bad(e) : good(x)
        )
      );
    } catch (e) {
      bad(e);
    }
    return () => {
    };
  })
);
var readDir = readDirWithConfig({});
var writeFileWithConfig = curry2(
  (conf, file, content) => new Future2((bad, good) => {
    __writeFile(file, content, conf, (e) => {
      if (e) {
        bad(e);
        return;
      }
      good(content);
    });
    return () => {
    };
  })
);
var writeFile = writeFileWithConfig({ encoding: "utf8" });
var removeFileWithConfig = curry2(
  (options, fd) => new Future2((bad, good) => {
    __rm(fd, options, (err) => err ? bad(err) : good(fd));
    return () => {
    };
  })
);
var DEFAULT_REMOVAL_CONFIG = {
  force: false,
  maxRetries: 0,
  recursive: false,
  retryDelay: 100,
  parallel: 10
};
var removeFile = removeFileWithConfig(DEFAULT_REMOVAL_CONFIG);
var removeFilesWithConfig = curry2(
  (conf, list) => pipe2(
    map(removeFileWithConfig(without(["parallel"], conf))),
    parallel(propOr2(10, "parallel", conf))
  )(list)
);
var removeFiles = removeFilesWithConfig(DEFAULT_REMOVAL_CONFIG);
var mkdir = curry2(
  (conf, x) => new Future2((bad, good) => {
    __mkdir(x, conf, (err) => err ? bad(err) : good(x));
    return () => {
    };
  })
);
var mkdirp = mkdir({ recursive: true });
var access = curry2(
  (permissions, filePath) => new Future2((bad, good) => {
    __access(filePath, permissions, (err) => err ? bad(err) : good(true));
    return () => {
    };
  })
);
var exists = access(constants.F_OK);
var directoryOnly = (filePath) => filePath.slice(0, filePath.lastIndexOf("/"));
var writeFileWithAutoPath = curry2(
  (filePath, content) => pipe2(
    directoryOnly,
    (dir) => pipe2(
      exists,
      chainRej(() => mkdirp(dir))
    )(dir),
    chain(() => writeFile(filePath, content))
  )(filePath)
);
var rm = curry2(
  (conf, x) => new Future2((bad, good) => {
    __rm(x, conf, (err) => err ? bad(err) : good(x));
    return () => {
    };
  })
);
var rimraf = rm({ force: true, recursive: true });

// src/path.js
import { join, normalize } from "node:path";
import { curry as curry3 } from "ramda";
var pathRelativeTo = curry3((pwd, x) => {
  if (typeof pwd !== "string" || typeof x !== "string") {
    throw new Error("Cannot normalize bad paths.");
  }
  return join(pwd, normalize(x));
});

// src/interpret.js
import { Future as Future3 } from "fluture";
var interpret = (filepath) => Future3((bad, good) => {
  import(filepath).catch(bad).then(good);
  return () => {
  };
});
export {
  DEFAULT_REMOVAL_CONFIG,
  access,
  directoryOnly,
  exists,
  flexeca,
  flexecaWithCanceller,
  interpret,
  localize,
  mkdir,
  mkdirp,
  pathRelativeTo,
  readDir,
  readDirWithConfig,
  readFile,
  readJSONFile,
  removeFile,
  removeFileWithConfig,
  removeFiles,
  removeFilesWithConfig,
  rimraf,
  rm,
  writeFile,
  writeFileWithAutoPath,
  writeFileWithConfig
};
