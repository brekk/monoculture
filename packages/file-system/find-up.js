import { Future } from 'fluture'
import { curry } from 'ramda'
import { findUp as __findUp } from 'find-up'

export const findUpWithCancel = curry((cancel, opts, x) =>
  Future((bad, good) => {
    __findUp(x, opts)
      .catch(bad)
      .then(raw => (raw ? good(raw) : bad(new Error('No config file found!'))))
    return cancel
  })
)
export const findUp = findUpWithCancel(() => {})
