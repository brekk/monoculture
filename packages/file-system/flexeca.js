import { pipe, curry, ifElse, propOr } from 'ramda'
import { execa } from 'execa'
import { Future } from 'fluture'

const didFail = propOr(true, 'failed')
const fail = propOr('Something broke', 'stderr')

export const flexecaWithOptionsAndCancel = curry(
  (cancellation, cmd, opts, args) =>
    new Future((bad, good) => {
      execa(cmd, args, opts)
        .catch(pipe(fail, z => `FAILED: "${z}"`, bad))
        .then(ifElse(didFail, pipe(fail, bad), good))
      return cancellation
    })
)

export const flexecaWithCanceller = curry(
  (cancellation, cmd, args) =>
    new Future((bad, good) => {
      execa(cmd, args)
        .catch(pipe(fail, z => `FAILED: "${z}"`, bad))
        .then(ifElse(didFail, pipe(fail, bad), good))
      return cancellation
    })
)

export const flexeca = flexecaWithCanceller(() => {})
