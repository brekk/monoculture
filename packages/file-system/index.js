export * from './find-up'
export * from './exec'
export * from './fs'
export * from './interpret'
export * from './path'
