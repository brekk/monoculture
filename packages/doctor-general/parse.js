import { basename, extname } from 'node:path'

import {
  unless,
  always as K,
  defaultTo,
  join,
  curry,
  filter,
  flatten,
  head,
  identity,
  last,
  map,
  pathOr,
  pipe,
  propOr,
  uniq,
} from 'ramda'
import { readFile } from 'file-system'
import { isJSDocComment, addLineNumbers, groupContiguousBlocks } from './file'
import { stripRelative } from './text'
import { nthIndex, lines } from 'knot'
import { objectifyComments } from './comment'

const getAny = curry((def, keyPath, comments) =>
  pipe(
    map(pathOr(def, keyPath)),
    filter(identity),
    uniq,
    x => x.sort(),
    head
  )(comments)
)

const getPageSummary = pipe(
  getAny('', ['structure', 'pageSummary']),
  defaultTo([]),
  join(' ')
)

const getPageTitle = getAny('', ['structure', 'page'])

const getPackage = i => {
  if (i.indexOf('/') > -1) {
    const y = nthIndex('/', -2, i)
    return y.slice(0, y.indexOf('/'))
  }
  return i
}

export const parse = curry((root, filename, content) => {
  const newName = stripRelative(filename)
  const slugName = basename(newName, extname(newName))
  return pipe(
    // String
    lines,
    // List String
    raw =>
      pipe(
        addLineNumbers,
        // List #[Integer, String]
        filter(pipe(last, isJSDocComment)),
        // List #[Integer, String]
        groupContiguousBlocks,
        // List #[Integer, String]
        objectifyComments(newName, raw),
        // List CommentBlock
        comments => ({
          slugName,
          package: getPackage(newName),
          pageTitle: getPageTitle(comments),
          pageSummary: getPageSummary(comments),
          filename: newName,
          comments,
          order: pipe(getAny('0', ['structure', 'order']), x =>
            parseInt(x, 10)
          )(comments),
          fileGroup: getAny('', ['fileGroup'], comments),
          links: pipe(map(propOr([], 'links')), flatten)(comments),
        })
      )(raw)
    // CommentedFile
  )(content)
})

// comments in pipe show current shape per step
export const parseFile = curry((debugMode, root, filename) =>
  pipe(
    // String
    readFile,
    // Future<Error, String>
    map(parse(root, filename)),
    // remove orphan comments (parser found it but its not well-formed)
    map(p => ({
      ...p,
      comments: pipe(
        filter(
          ({ lines: l, start, end, summary }) =>
            start !== end && !!summary && l.length > 0
        ),
        unless(
          K(debugMode),
          map(({ lines: _lines, ...rest }) => rest)
        )
      )(p.comments),
    }))
    // CommentedFile
  )(filename)
)
