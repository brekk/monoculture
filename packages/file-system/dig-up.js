import { Future } from 'fluture'
import { curry } from 'ramda'
import { findUp as __findUp } from 'find-up'

export const digUpWithCancel = curry(
  function _digUpWithCancel(cancel, opts, x) {
    return Future((bad, good) => {
      __findUp(x, opts)
        .catch(bad)
        .then(raw =>
          raw ? good(raw) : bad(new Error('No config file found!'))
        )
      return cancel
    })
  }
)
export const digUp = digUpWithCancel(() => {})
