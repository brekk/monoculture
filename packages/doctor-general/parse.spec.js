import { fork } from 'fluture'
import { relativePathJoin } from 'file-system'

import { parse, parseFile } from './parse'

test('parse', () => {
  const filename = 'i-yam-what-i.yaml'
  const content = `
/**
 * Very cool
 * @name cool
 * @private
 * @see {@link emoji.sunglasses}
 * @example
 * \`\`\`ts
 * import { cool } from './parse'
 * \`\`\`
 */

the content here will be thrown right in the trash

/**
 * Nice
 * @name nice
 * @example
 * \`\`\`ts
 * import { nice } from './parse'
 * \`\`\`
 */
`
  expect(parse('root', filename, content)).toEqual({
    comments: [
      {
        addTo: '',
        end: 10,
        fileGroup: '',
        keywords: ['@example', '@link', '@name', '@private', '@see'],
        lines: [
          'Very cool',
          '@name cool',
          '@private',
          '@see {@link emoji.sunglasses}',
          '@example',
          '```ts',
          "import { cool } from './parse'",
          '```',
        ],
        links: ['emoji.sunglasses'],
        start: 1,
        structure: {
          example: `\`\`\`ts
import { cool } from './parse'
\`\`\``,
          name: 'cool',
          private: true,
          see: ['emoji.sunglasses'],
        },
        summary: 'Very cool',
      },
      {
        addTo: '',
        end: 21,
        fileGroup: '',
        keywords: ['@example', '@name'],
        lines: [
          'Nice',
          '@name nice',
          '@example',
          '```ts',
          "import { nice } from './parse'",
          '```',
        ],
        links: [],
        start: 14,
        structure: {
          example: `\`\`\`ts
import { nice } from './parse'
\`\`\``,
          name: 'nice',
        },
        summary: 'Nice',
      },
    ],
    fileGroup: undefined,
    filename: 'i-yam-what-i.yaml',
    links: ['emoji.sunglasses'],
    order: 0,
    package: 'i-yam-what-i.yaml',
    pageSummary: '',
    pageTitle: undefined,
    slugName: 'i-yam-what-i',
  })
})
test('parseFile', done => {
  const input = relativePathJoin(__dirname, './file.js')
  fork(done)(raw => {
    const cleanFilename = raw.filename.split('/').slice(-3).join('/')
    expect({ ...raw, filename: cleanFilename }).toMatchSnapshot()
    done()
  })(parseFile(false, 'root', input))
})
