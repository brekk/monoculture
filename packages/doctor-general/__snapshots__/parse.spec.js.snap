// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseFile 1`] = `
{
  "comments": [
    {
      "addTo": "",
      "end": 77,
      "fileGroup": "",
      "keywords": [
        "@example",
        "@exported",
        "@name",
        "@param",
        "@returns",
        "@signature",
      ],
      "links": [],
      "start": 49,
      "structure": {
        "example": "\`\`\`js
const a = { a: true, greeting: 'hello', comments: ['one', 'two'], links: ['a', 'b'] }
const b = { b: true, greeting: 'ahoy', comments: ['three', 'four'], links: ['c', 'd'] }
expect(combineFiles(true, a, b)).toEqual({
  a: true,
  b: true,
  greeting: 'ahoy',
  comments: ['one', 'two', 'three', 'four'],
  links: ['a', 'b', 'c', 'd']
})
expect(combineFiles(false, a, b)).toEqual({
  a: true,
  b: true,
  greeting: 'hello',
  comments: ['three', 'four', 'one', 'two'],
  links: ['c', 'd', 'a', 'b']
})
\`\`\`",
        "exported": true,
        "name": "combineFiles",
        "param": [
          [
            "{boolean}",
            "leftToRight",
            "Associate",
            "left",
            "to",
            "right",
          ],
          [
            "{File}",
            "a",
            "left",
            "file",
          ],
          [
            "{File}",
            "b",
            "right",
            "file",
          ],
        ],
        "returns": [
          "{File}",
          "Merged",
          "file",
        ],
        "signature": [
          "boolean",
          "->",
          "File",
          "->",
          "File",
          "->",
          "File",
        ],
      },
      "summary": "Merge two file representations. Can be right or left associative",
    },
  ],
  "fileGroup": undefined,
  "filename": "packages/doctor-general/file.js",
  "links": [],
  "order": 0,
  "package": "doctor-general",
  "pageSummary": "",
  "pageTitle": undefined,
  "slugName": "file",
}
`;
